File: components/api/api-modules-python/src/main/resources/META-INF/dirigible/python-modules/dirigible/io/streams.py
# Copyright (c) 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
# All rights reserved. This program and the accompanying materials
# are made available under the terms of the Eclipse Public License v2.0
# which accompanies this distribution, and is available at
# http://www.eclipse.org/legal/epl-v20.html
# SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
# SPDX-License-Identifier: EPL-2.0

import bytes
import java

StreamsFacade = java.type("org.eclipse.dirigible.components.api.io.StreamsFacade")

class InputStream:
    @staticmethod
    def __init__(self, native):
        self.native = native

    @staticmethod
    def read(self):
        return StreamsFacade.read(self.native)

    @staticmethod
    def readBytes(self):
        native = StreamsFacade.readBytes(self.native)
        return bytes.toPythonBytes(native)

    @staticmethod
    def readBytesNative(self):
        return StreamsFacade.readBytes(self.native)

    @staticmethod
    def readText(self):
        return StreamsFacade.readText(self.native)

    @staticmethod
    def close(self):
        StreamsFacade.close(self.native)

    @staticmethod
    def isValid(self):
        return self.native is not None

class OutputStream:
    @staticmethod
    def __init__(self, native):
        self.native = native

    @staticmethod
    def write(self, byte):
        StreamsFacade.write(self.native, byte)

    @staticmethod
    def writeBytes(self, data):
        native = bytes.toJavaBytes(data)
        StreamsFacade.writeBytes(self.native, native)

    @staticmethod
    def writeBytesNative(self, data):
        StreamsFacade.writeBytes(self.native, data)

    @staticmethod
    def writeText(self, text):
        StreamsFacade.writeText(self.native, text)

    @staticmethod
    def close(self):
        StreamsFacade.close(self.native)

    @staticmethod
    def getBytes(self):
        native = StreamsFacade.getBytes(self.native)
        data = bytes.toPythonBytes(native)
        return data

    @staticmethod
    def getBytesNative(self):
        native = StreamsFacade.getBytes(self.native)
        return native

    @staticmethod
    def getText(self):
        value = StreamsFacade.getText(self.native)
        return value

    @staticmethod
    def isValid(self):
        return self.native is not None

def copy(input, output):
    StreamsFacade.copy(input.native, output.native)

def copyLarge(input, output):
    StreamsFacade.copyLarge(input.native, output.native)

def getResourceAsByteArrayInputStream(path):
    native = StreamsFacade.getResourceAsByteArrayInputStream(path)
    return InputStream(native)

def createByteArrayInputStream(data):
    array = bytes.toJavaBytes(data)
    native = StreamsFacade.createByteArrayInputStream(array)
    return InputStream(native)

def createByteArrayOutputStream():
    native = StreamsFacade.createByteArrayOutputStream()
    return OutputStream(native)

def createInputStream(native):
    inputStream = InputStream(native)
    return inputStream

def createOutputStream(native):
    outputStream = OutputStream(native)
    return outputStream

File: components/api/api-modules-javascript/src/main/resources/META-INF/dirigible/modules/src/io/streams.ts
/*
 * Copyright (c) 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 * SPDX-License-Identifier: EPL-2.0
 */
import { Bytes } from "sdk/io/bytes";

const StreamsFacade = Java.type("org.eclipse.dirigible.components.api.io.StreamsFacade");

/**
 * InputStream object. To be used internally by the API layer
 */
export class InputStream {

	public readonly native: any;

	constructor(native: any) {
		this.native = native;
	}

	public read(): number {
		return StreamsFacade.read(this.native);
	}

	public readBytes(): any[] {
		const native = StreamsFacade.readBytes(this.native);
		return Bytes.toJavaScriptBytes(native);
	}

	public readBytesNative(): any[] {
		return StreamsFacade.readBytes(this.native);
	}

	public readText(): string {
		return StreamsFacade.readText(this.native);
	}

	public close(): void {
		StreamsFacade.close(this.native);
	}

	public isValid(): boolean {
		return this.native !== undefined && this.native !== null;
	}
}

/**
 * OutputStream object. To be used internally by the API layer
 */
export class OutputStream {

	public readonly native: any;

	constructor(native: any) {
		this.native = native;
	}

	public write(byte: number): void {
		StreamsFacade.write(this.native, byte);
	}

	public writeBytes(data: any[]): void {
		const native = Bytes.toJavaBytes(data);
		StreamsFacade.writeBytes(this.native, native);
	}

	public writeBytesNative(data: any[]): void {
		StreamsFacade.writeBytes(this.native, data);
	}

	public writeText(text: string): void {
		StreamsFacade.writeText(this.native, text);
	}

	public close(): void {
		StreamsFacade.close(this.native);
	}

	public getBytes(): any[] {
		const native = StreamsFacade.getBytes(this.native);
		return Bytes.toJavaScriptBytes(native);
	}

	public getBytesNative(): any[] {
		return StreamsFacade.getBytes(this.native);
	}

	public getText(): string {
		return StreamsFacade.getText(this.native);
	}

	public isValid(): boolean {
		return this.native !== undefined && this.native !== null;
	}
}

export class Streams {

	public static copy(input: InputStream, output: OutputStream): void {
		StreamsFacade.copy(input.native, output.native);
	}
	
	public static copyLarge(input: InputStream, output: OutputStream) {
		StreamsFacade.copyLarge(input.native, output.native);
	}

	/**
	 * Get an ByteArrayInputStream for the provided resource
	 */
	public static getResourceAsByteArrayInputStream(path: string): InputStream {
		const native = StreamsFacade.getResourceAsByteArrayInputStream(path);
		return new InputStream(native);
	}

	/**
	 * Create an ByteArrayInputStream for byte array provided
	 */
	public static createByteArrayInputStream(data: any[]): InputStream {
		const array = Bytes.toJavaBytes(data);
		const native = StreamsFacade.createByteArrayInputStream(array);
		return new InputStream(native);
	}

	/**
	 * Create a ByteArrayOutputStream
	 */
	public static createByteArrayOutputStream(): OutputStream {
		const native = StreamsFacade.createByteArrayOutputStream();
		return new OutputStream(native);
	}

	/**
	 * Create an InputStream object by a native InputStream
	 */
	public static createInputStream(native: any) {
		const inputStream = new InputStream(native);
		return inputStream;
	}

	/**
	 * Create an OutputStream object by a native OutputStream
	 */
	public createOutputStream(native: any) {
		const outputStream = new OutputStream(native);
		return outputStream;
	}
}

// @ts-ignore
if (typeof module !== 'undefined') {
	// @ts-ignore
	module.exports = Streams;
}
